source/material_model/dynamic_friction.cc

```diff
@@ -0,0 +1,444 @@
+            prm.declare_entry ("Maximum viscosity", "1e24",
+                               Patterns::Double (0),
+                               "The value of the maximum viscosity cutoff $\\eta_max$. Units: $Pa\\;s$.");
+            prm.declare_entry ("Reference strain rate", "1e-15",
+                               Patterns::Double (0),
+                               "The value of the initial strain rate prescribed during the "
+                               "first nonlinear iteration $\\dot{\\epsilon}_ref$. Units: $1/s$.");
+            prm.declare_entry ("Coefficients of static friction", "0.5",
+                             Patterns::List(Patterns::Double(0)),
+                             "List of coefficients of static friction for background mantle and compositional fields,"
+                             "for a total of N+1 values, where N is the number of compositional fields."
+                             "If only one value is given, then all use the same value. Units: $dimensionless$");
+            prm.declare_entry ("Coefficients of dynamic friction", "0.4",
+                             Patterns::List(Patterns::Double(0)),
+                             "List of coefficients of dynamic friction for background mantle and compositional fields,"
+                             "for a total of N+1 values, where N is the number of compositional fields."
+                             "If only one value is given, then all use the same value. Units: $dimensionless$");
+            prm.declare_entry ("Cohesions", "4.e6",
+                             Patterns::List(Patterns::Double(0)),
+                             "List of cohesions for background mantle and compositional fields,"
+                             "for a total of N+1 values, where N is the number of compositional fields."
+                             "If only one value is given, then all use the same value. Units: $Pa$");
+            prm.declare_entry ("Background Viscosities", "1.e20",
+                             Patterns::List(Patterns::Double(0)),
+                             "List of background viscosities for mantle and compositional fields,"
+                             "for a total of N+1 values, where N is the number of compositional fields."
+                             "If only one value is given, then all use the same value. Units: $Pa s $");                    
+          }
+          prm.leave_subsection();
+        }
+        prm.leave_subsection();
+      }
+      prm.leave_subsection();
+    }
+
+    template <int dim>
+    void
+    DynamicFriction<dim>::parse_parameters (ParameterHandler &prm)
+    {
+      //not pretty, but we need to get the number of compositional fields before
+      //simulatoraccess has been initialized here...
+      unsigned int n_foreground_fields;
+      prm.enter_subsection ("Compositional fields");
+      {
+        n_foreground_fields = prm.get_integer ("Number of fields");
+      }
+      prm.leave_subsection();
```

Comment:
You can remove everything above this line in this function, and then instead of `n_foreground_fields` use `this->n_compositional_fields()`. This code is a leftover from years ago.